src/LayerPlanBuffer.cpp

```diff
@@ -162,16 +175,149 @@ void LayerPlanBuffer::insertPreheatCommand(std::vector<ExtruderPlan*>& extruder_
+        insertFinalPrintTempCommand(extruder_plans, extruder_plan_idx - 1);
+        insertPrintTempCommand(extruder_plan);
+    }
+}
+
+void LayerPlanBuffer::insertPrintTempCommand(ExtruderPlan& extruder_plan)
+{
+    unsigned int extruder = extruder_plan.extruder;
+    double print_temp = extruder_plan.printing_temperature;
+
+    double heated_pre_travel_time = 0;
+    if (preheat_config.getInitialPrintTemp(extruder) != 0)
+    { // handle heating from initial_print_temperature to printing_tempreature
+        unsigned int path_idx;
+        for (path_idx = 0; path_idx < extruder_plan.paths.size(); path_idx++)
+        {
+            GCodePath& path = extruder_plan.paths[path_idx];
+            heated_pre_travel_time += path.estimates.getTotalTime();
+            if (!path.isTravelPath())
+            {
+                break;
+            }
+        }
+        bool wait = false;
+        extruder_plan.insertCommand(path_idx, extruder, print_temp, wait);
+    }
+    extruder_plan.heated_pre_travel_time = heated_pre_travel_time;
+}
+
+void LayerPlanBuffer::insertFinalPrintTempCommand(std::vector<ExtruderPlan*>& extruder_plans, unsigned int extruder_plan_idx)
+{
+    ExtruderPlan& last_extruder_plan = *extruder_plans[extruder_plan_idx];
+    int extruder = last_extruder_plan.extruder;
+
+    double final_print_temp = preheat_config.getFinalPrintTemp(extruder);
+    if (final_print_temp == 0)
+    {
+        return;
+    }
+
+    double heated_post_travel_time = 0;
+    { // compute heated_post_travel_time
+        unsigned int path_idx;
+        for (path_idx = last_extruder_plan.paths.size() - 1; int(path_idx) >= 0; path_idx--)
+        {
+            GCodePath& path = last_extruder_plan.paths[path_idx];
+            if (!path.isTravelPath())
+            {
+                break;
+            }
+            heated_post_travel_time += path.estimates.getTotalTime();
+        }
+    }
+
+    double time_window = 0;
+    double average_print_temp = 0;
+    double initial_print_temp = -1;
+    { // compute time window and print temp statistics
+        double heated_pre_travel_time;
+        unsigned int print_temp_count = 0;
+        for (unsigned int prev_extruder_plan_idx = extruder_plan_idx; (int)prev_extruder_plan_idx >= 0; prev_extruder_plan_idx--)
+        {
+            ExtruderPlan& prev_extruder_plan = *extruder_plans[prev_extruder_plan_idx];
+            if (prev_extruder_plan.extruder != extruder)
+            {
+                break;
+            }
+            time_window += prev_extruder_plan.estimates.getTotalTime();
+            heated_pre_travel_time = prev_extruder_plan.heated_pre_travel_time;
+
+            if (prev_extruder_plan.estimates.getTotalUnretractedTime() > 0 && prev_extruder_plan.estimates.getMaterial() > 0)
+            { // handle temp statistics
+                assert(prev_extruder_plan.printing_temperature != -1 && "Previous extruder plan should already have a temperature planned");
```

Comment:
-1 is a stub which is used as invalid temperature by programmers.
0 is a stub which is used as 'don't use this temp variable' by users.
So: -1 is a programmer special case, while 0 is a special case which could be initiated by the user.

Should I use -1337 as invalid temp? Then it's not a valid temp anyhow...

Do you think I should introduce extra settings for the 0-option?
I don't like introducing too many settings, but I could introduce a 'Use Initial Printing Temperature' and a 'Use Final Printing Temperature'...